### DISCUSSION TITLE
Add null-forgiving operator for nullable value types

### DISCUSSION THREAD

**Discussion by cytoph** [dotnet/csharplang#5412] [Language Ideas] [answered]
### C# 6 - null-conditional operator
We now have the [? [null-conditional] operators](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-) since C# 6, with them we could do the following to let a member or element operation apply to its operand only if that operand evaluates to non-null.
```
A?.B?.Do(C);
A?.B?.[C];
```

### C# 8 - null-forgiving operator (for nullable reference types only)
Then, with C# 8, we got [nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-reference-types) and with that the [! [null-forgiving] operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) which (of course only in an enabled [nullable annotation context](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references#nullable-contexts)) can be used to declare that an expression of a reference type isn't `null`. Like this:
```
public static void Main()
{
    Person? p = Find("John");
    if (IsValid(p))
    {
        Console.WriteLine($"Found {p!.Name}");
    }
}

public static bool IsValid(Person? person)
    => person is not null && person.Name is not null;
```

### Proposal - null-forgiving operator (for nullable value types, enumeration types, etc.)
But what still is missing is the ability to ignore/suppress/forgive `null` for [nullable value types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types) or [enumeration types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/enum) and such, like this:
```
public struct HexColor
{
    public byte Red { get; set; }
    public byte Green { get; set; }
    public byte Blue { get; set; }
}

public static void Main()
{
    HexColor? color = MaybeGetAColor();

    if (color.HasValue)
    {
        // do something else

        Console.WriteLine($"Red content: {color!.Red}");
        Console.WriteLine($"Green content: {color!.Green}");
        Console.WriteLine($"Blue content: {color!.Blue}");
    }
}
```
Of course, this is only a shorter a shorter version of (in the previous example) `color.Value.Red`, but so is `color?.Red` for `(color != null ? color.Red : null)`, and I think many language proposals are here for being able to get more done with less code.

**Comment by svick** [comment, marked as answer]
This is a duplicate of https://github.com/dotnet/csharplang/discussions/841.

Also note that your example could be written today like this:

```c#
if (MaybeGetAColor() is {} color)
{
    Console.WriteLine($"Red content: {color.Red}");
    Console.WriteLine($"Green content: {color.Green}");
    Console.WriteLine($"Blue content: {color.Blue}");
}
```

**Reply by cytoph** [reply]
Yeah, for the example above that would be option, I know.
But so would it be for not using null-coalescing.
So if null-coalescing is a thing, null-forgiving is just as much justified.

And if at all it would be a duplicate of the [comment](https://github.com/dotnet/csharplang/issues/1865#issuecomment-544685799)  in #1865 which is stated in the last comment of the issue, you wrote, but I don't think any of those described the problem and especially the justification of such a features as much (also they are both pretty old).

**Reply by 333fred** [reply]
LDM explicitly considered and rejected 841. The nullable reference types feature was designed to _not_ change runtime behavior of code: that includes the `!` operator, and includes features of NRT applied to nullable value types. The consequence of this is that `nullableValue!` already means something today, and it would be a breaking change to change that.
